include/Transforms/Functional.h

```diff
@@ -0,0 +1,364 @@
+  // Save a copy of the scope.
+  SmallVector<std::unique_ptr<Region>> backup = cloneRegionsOf(scope);
+
+  // FIXME: The sequence needs to set insertion points itself. This should be
+  // done dynamically in PDL.
+  auto result = std::forward<SequenceT>(sequence)(op, rewriter,
+                                                  std::forward<Args>(args)...);
+  // If the match failed, rewind the scope.
+  if (failed(result))
+    takeRegions(scope, backup);
+  return result;
+}
+} // end namespace detail
+
+/// Apply a pattern sequence directly on an operation. If any pattern in the
+/// sequence fails, then the state of the IR is restored. Patterns in the
+/// sequence should not query or modify IR outside the nearest scope.
+template <typename OpT, typename SequenceT, typename... Args>
+auto applySequenceAt(OpT op, SequenceT &&sequence, Args &&...args) {
+  Operation *scope = findNearestScope(op);
+  assert(scope && "could not find a valid scope");
+  detail::SimpleRewriter rewriter(scope->getContext());
+  return detail::applySequence(rewriter, scope, op,
+                               std::forward<SequenceT>(sequence),
+                               std::forward<Args>(args)...);
+}
+
+namespace detail {
+template <typename OpT>
+OpT walkUntil(Operation *scope, unsigned walkIdx) {
+  unsigned curIdx = 0;
+  OpT result = nullptr;
+  scope->walk([&](OpT op) {
+    if (curIdx++ < walkIdx)
+      return WalkResult::advance();
+    result = op;
+    return WalkResult::interrupt();
+  });
+  return result;
+}
+} // end namespace detail
+
+template <typename SequenceT, typename... Args>
+auto applySequenceOnce(Operation *scope, SequenceT &&sequence, Args &&...args) {
+  assert(scope->hasTrait<OpTrait::IsIsolatedFromAbove>() &&
+         "scope is not isolated from above");
+  using Traits = llvm::function_traits<std::decay_t<SequenceT>>;
+  using OpT = typename Traits::template arg_t<0>;
+
+  detail::SimpleRewriter rewriter(scope->getContext());
+
+  unsigned walkIdx = 0;
+  // FIXME: We have to restart the walk each time because the iteration gets
+  // invalidated on failures.
+  for (OpT op = nullptr; (op = detail::walkUntil<OpT>(scope, walkIdx));
+       ++walkIdx) {
+    SmallVector<std::unique_ptr<Region>> backup = cloneRegionsOf(scope);
+    auto result = std::forward<SequenceT>(sequence)(
+        op, rewriter, std::forward<Args>(args)...);
+    // If the pattern succeeded, discard the backup and return the match.
+    if (succeeded(result))
+      return result;
+    // Otherwise, rewind the scope.
+    takeRegions(scope, backup);
+  }
+  return typename Traits::result_t();
+}
+
+using SequencePattern =
+    llvm::unique_function<LogicalResult(Operation *, PatternRewriter &)>;
+
+namespace impl {
+void applyPatternsRecursively(Operation *scope,
+                              MutableArrayRef<SequencePattern> patterns);
+} // end namespace impl
+
+namespace detail {
+template <typename OpT>
+struct IsaOr {
```

Comment:
I do not understand what this does :)

Reply:
Let me document this in a little bit more detail...